### DISCUSSION TITLE
Chooser Related sub-Solutions

### DISCUSSION THREAD

**Discussion by hazho** [wagtail/wagtail#8514] [General]
this discussion is for the core work contribution on the Chooser panels and modals, Goals are:
- Enhancing the handlers of activities done in modal.
- Enhancing UX.
- Enhancing Security.
- Enhancing Performance.
- Providing a generic UI that can be modified by Developers.

Contributions are very welcomed..!

**Comment by hazho** [comment]
Axios uses this API as a base proxy for things like Chooser Panels/Modals
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises
I will try to explore more and analyze it to reache the best possible approach, if anyone has an experience with axios and/or Promises API, please share your thoughts

**Comment by gasman** [comment]
I've been doing a lot of work this week on eliminating duplicated code across choosers, with the aim of both cleaning up the codebase and making it possible to define new choosers for arbitrary models with just a few lines of code. The master plan is outlined in [#8480 - Generic chooser views](https://github.com/wagtail/wagtail/issues/8480).

The chooser code spans a lot of areas - the chooser views (which themselves have various subcomponents - search, filtering by collection, 'create new' tabs), the client-side code to manage displaying them as modals, the chooser form widgets, and StreamField blocks - which can be tackled as largely separate cleanup tasks. [#8501 - chooser widget cleanup](https://github.com/wagtail/wagtail/pull/8501) is now in review and [#8532 - chooser modal JS cleanup](https://github.com/wagtail/wagtail/pull/8532) is a draft PR - other code is not quite ready for PRs yet, but should be soon.

The current state of play is that a set of completely standard chooser views with no customisations can be set up with just a few lines of configuration that look like this:

```python
from wagtail.admin.viewsets.chooser import ChooserViewSet

from bakerydemo.base.models import People


class PersonChooserViewSet(ChooserViewSet):
    model = People
    page_title = "Choose a person"
    icon = "user"
```

In practice, most choosers will need some amount of customisation (all of the existing built-in ones do), but the hope is that this will continue to be done mainly through configuration rather than custom code. So, for example, if we want our person chooser listing to have distinct columns for first name and last name, it'll be something like `columns = ['first_name', 'last_name']` rather than having to override templates.

Meanwhile, with #8501 in place the code for a chooser widget looks like this:

```python
import json

from django.utils.translation import gettext_lazy as _

from wagtail.admin.widgets.chooser import BaseChooser

from bakerydemo.base.models import People


class PersonChooser(BaseChooser):
    choose_one_text = _("Choose a person")
    choose_another_text = _("Choose another person")
    link_to_chosen_text = _("Edit this person")

    chooser_modal_url_name = "person_chooser:choose"
    icon = "user"
    model = People

    def render_js_init(self, id_, name, value_data):
        return "new PersonChooser({0});".format(json.dumps(id_))

    class Media:
        js = [
            "personchooser/js/person-chooser-modal.js",
            "personchooser/js/person-chooser.js",
        ]
```

I think we can do better than that. The `chooser_modal_url_name`, `icon` and `model` fields are duplicating information that PersonChooserViewSet already has - my plan is that the ChooserViewSet object will provide a preconfigured widget class as an extra feature (or perhaps there'll be some kind of über-object that defines the viewset and widget in one go, just like the viewset defines a collection of views).

Once nice feature of the new implementation is that the "edit this object" link is built using the [AdminURLFinder](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/admin_url_finder.py) mechanism that we put in place for the audit logging feature in 2.15, so that the chooser doesn't have to have built-in knowledge of where the corresponding edit view is (snippet, modeladmin, or a dedicated app).

The Javascript is turning out to be one of the trickiest bits to refactor, since it's not written in a very object-oriented style - it's based around passing bundles of event handler callbacks around, so the strategy of defining a base class and overriding it with special cases isn't available (and the existing logic is baked deeply enough into the Draftail integration and third-party code that replacing it outright isn't a feasible option). However, in #8532 I've been able to hack out large parts of the duplicated code and move it into a utility module, including a new SearchController class which handles the common pattern of re-rendering the results panel in response to search queries and pagination. Hopefully when this bit is complete we'll eliminate the need for simple choosers to ship with their own JS file.

**Reply by hazho** [reply]
I liked the new PR, especially the:

```
    show_edit_link = True
    show_clear_link = True
```
and the hidden fields, I don't think is a necessary thing to separate form each others, hidden inputs are already hidden, it is an extra job to process the difference, instead, for all kind of form fields, a ONE loop is viable, am I wrong?

**Reply by hazho** [reply]
Regarding the modal itself, what other handlers needed, beside the choosen object?
if nothing else, this should be very simple to handle, with few line of ES6
is it needed to do some other handling for streamfield serialization?

**Reply by gasman** [reply]
> I don't think is a necessary thing to separate form each others, hidden inputs are already hidden, it is an extra job to process the difference, instead, for all kind of form fields, a ONE loop is viable, am I wrong?

See https://github.com/wagtail/wagtail/blob/main/wagtail/admin/templates/wagtailadmin/generic/form.html for an example of how this is used. Visible fields will be rendered with labels, padding and other bits of page furniture - we don't want to show those for hidden fields.

**Comment by hazho** [comment]
Regarding the `chooser_modal_url_name` Why not getting the common listing URL of the target model?
this way, we can get rid of the extra processing and also unite the permissions (especially object-based permissions), for example:
for page Chooser modal, the returned template would be a minimized version of the page explorer listing, and for Image Chooser, a minimized version of the Image listing template, hense the views to controll the permissions, this can solve many other related problem that need to be solved for multisite purposes as well

**Reply by gasman** [reply]
Visually the chooser interface is quite similar to the index listing, but the logic behind it is very different. The links need to point to different places; the permission checks are different (a user needs to be able to choose a page that they don't have edit permission for), and the action menu (move, copy, delete...) shouldn't be shown within the chooser. A long time back [these listings used the same template](https://github.com/wagtail/wagtail/blob/stable/0.6.x/wagtail/wagtailadmin/templates/wagtailadmin/pages/list.html) - it was a total mess, since you needed `{% if choosing %}` blocks all over the place.

**Reply by hazho** [reply]
Can't we make a generic template for the listing, and then `{% include '<the page exploring index> or <page chooser index for linking> or <page chooser index for copying/moving>' %} `?
permissions for included template-components would be handled by viewsets as how the template_tags similarly works

**Comment by hazho** [comment]
Just for record, it turns out that using **async/await** is better than using **Promises**  for many reasons, especially the performance
